s2spy/time.py

```diff
@@ -304,3 +296,237 @@ def show(self) -> pd.DataFrame:
+            # count length
+            self._total_length_precursor += period_block.length + period_block.gap
+
+    def _append_check(self, list_period, period_block):
+        """Rules check for appending target precursor blocks."""
+        if not list_period:
+            if period_block.gap < 0:
+                raise ValueError(
+                    "First appended period block should not contain a negative gap."
+                )
+        else:
+            if (period_block.gap + list_period[-1].length) < 0:
+                raise ValueError(
+                    "Appended period block has negative gap larger than"
+                    + " the length of precedent period block."
+                )
+
+    def _map_year(self, year: int):
+        """Replace old map_year function"""
+        # flag the calendar as existed
+        self._end_indentifier = True
+
+        intervals_target = self._concatenate_periods(
+            year, self._list_target, self._total_length_target, True
+        )
+        intervals_precursor = self._concatenate_periods(
+            year, self._list_precursor, self._total_length_precursor, False
+        )
+
+        self.n_targets = len(intervals_target)
+        year_intervals = intervals_precursor[::-1] + intervals_target
+
+        # turn the list of intervals into pandas series
+        year_intervals = pd.Series(year_intervals[::-1], name=str(year))
+        year_intervals.index.name = "i_interval"
+        return year_intervals
+
+    def _concatenate_periods(self, year, list_periods, total_length, is_target):
+        # calculate start and end date
+        if is_target:
+            # anchor date is included
+            start_date = self._get_anchor(year)
+            end_date = start_date + pd.Timedelta(total_length, unit="D")
+            # generate date ranges
+            date_range = pd.date_range(start=start_date, end=end_date)
+        else:
+            # anchor date is excluded
+            start_date = self._get_anchor(year) - pd.Timedelta(
+                total_length + 1, unit="D"
+            )
+            end_date = self._get_anchor(year) - pd.Timedelta(1, unit="D")
+            # generate date ranges
+            date_range = pd.date_range(start=start_date, end=end_date)
+            date_range = date_range[::-1]
+        # generate intervals based on the building blocks
+        intervals = []
+        # pointer to the index of date for right boundary of interval
+        left_date_index = 0
+        # loop through all the building blocks to
+        for block in list_periods:
+            left_date_index += block.gap
+            # pointer to the index of date for left boundary of interval
+            right_date_index = left_date_index + block.length - 1
+            if is_target:
+                intervals.append(
+                    pd.Interval(
+                        date_range[left_date_index] - pd.Timedelta(1, unit="D"), # open left close right
```

Comment:
It is not a bug 😅. I have mentioned this in teams in our conversation. Perhaps you missed that comment. And my thought was partly the same as what you put here. We had closed right side before because of the `Adventcalendar`, which has the anchor date placed in the end. Now we have it in the middle. If we stick to the same "left open, right closed" rule, then the anchor date is invisible when you display the calendar. That is how the code works now.

I have considered to change it to "left closed, right open", since we move the anchor date in the middle. Given the conflict of two designs, we need to make a choice. Well, there are several things we should take into account:
- We expect users to use `AdventCalendar` and its variants more often
- This custom calendar will become the base calendar and it won't be shown to the users in most of the cases
- The `CustomCalendar` will only be shown to the advanced users

In other words, `AdventCalendar` has priority. So I decide to keep "left open, right closed" rule. That's why. I think we'd better have a discussion about it.

Reply:
What I have in mind now, is we make this `CustomCalendar` "left closed, right open", and when update `AdventCalendar` based on `CustomCalendar`, we change the results to be "left open, right closed". So we take care of both cases.

Reply:
> What I have in mind now, is we make this `CustomCalendar` "left closed, right open", and when update `AdventCalendar` based on `CustomCalendar`, we change the results to be "left open, right closed". So we take care of both cases.

Yes, I believe that as we're switching the anchor to be on the left side of the interval, the intervals _have_ to be left closed, right open. Refactoring the AdventCalendar/other calendars to this system can indeed come later.